File: pwa/sync.js
/* MyMoney PWA — motore di sincronizzazione (v2 Fase 4).
   Protocollo stile Cashew: diario locale + merge LWW + push/pull via API HTTP.
   Il contratto è in docs/SYNC-PROTOCOL.md.

   Flusso:
   1. Ogni operazione locale (aggiungi/modifica/elimina) scrive nel diario IndexedDB
   2. fullSync(): push → POST /api/finanze/ops (invia le ops locali non pushate)
                  pull → GET /api/finanze/diary?since=N (scarica ops del PC)
   3. Le ops remote vengono applicate in locale con merge LWW (stesse regole del PC) */
window.SYNC = (function () {
  "use strict";

  // ── device id ────────────────────────────────────────────────────────────
  function getDeviceId() {
    return DB.getMeta("device_id").then(function (id) {
      if (id) return id;
      var newId = "pwa_" + nuovoUid().substring(0, 12);
      return DB.setMeta("device_id", newId).then(function () { return newId; });
    });
  }

  // ── registrazione nel diario ─────────────────────────────────────────────
  function recordOp(entity, op, fields, deviceId) {
    var entry = {
      schema: 1,
      uid: fields.uid,
      entity: entity,
      op: op,
      fields: fields,
      rev: fields.rev || 1,
      updated_at: fields.updated_at || new Date().toISOString(),
      device_id: deviceId,
      ts: new Date().toISOString(),
      _pushed: false    // false = non ancora inviato al PC
    };
    return DB.appendDiary(entry);
  }

  // ── merge LWW ────────────────────────────────────────────────────────────
  function _wins(remoteRev, remoteUpdated, remoteDevice,
                 localRev, localUpdated, localDevice) {
    /* Vince il più alto: (rev, updated_at, device_id) */
    var rr = remoteRev || 0, lr = localRev || 0;
    if (rr !== lr) return rr > lr;
    var ru = String(remoteUpdated || ""), lu = String(localUpdated || "");
    if (ru !== lu) return ru > lu;
    return String(remoteDevice || "") > String(localDevice || "");
  }

  function _applyFields(existing, entity, fields, stores) {
    /* Sovrascrive i campi del record locale con quelli remoti */
    var obj = existing || {};
    obj.uid = fields.uid;
    obj.rev = fields.rev || 1;
    obj.updated_at = fields.updated_at;
    obj.deleted = !!fields.deleted;

    if (entity === "wallet") {
      obj.nome = fields.nome || "";
      obj.tipo = fields.tipo || "altro";
      obj.saldo_iniziale = fields.saldo_iniziale || 0;
      obj.note = fields.note || "";
      obj.ordine = fields.ordine || 0;
      obj.archiviato = !!fields.archiviato;
      obj.colore = fields.colore || "";
    } else if (entity === "category") {
      obj.nome = fields.nome || "";
      obj.kind = fields.kind || "";
      obj.archiviato = !!fields.archiviato;
    } else if (entity === "transaction") {
      obj.tipo = fields.tipo || "uscita";
      obj.data = fields.data;
      obj.importo = fields.importo || 0;
      obj.wallet_uid = fields.wallet_uid || null;
      obj.wallet_to_uid = fields.wallet_to_uid || null;
      obj.categoria_uid = fields.categoria_uid || null;
      obj.descrizione = fields.descrizione || "";
      obj.giro_id = fields.giro_id || "";
      obj.giro_aperta = !!fields.giro_aperta;
      obj.importo_ricevuto = fields.importo_ricevuto != null ? fields.importo_ricevuto : null;
      obj.data_ricevuto = fields.data_ricevuto || null;
      obj.controparte = fields.controparte || "";
    }
    return obj;
  }

  var ENTITY_STORE = { wallet: "wallets", category: "categorie", transaction: "movimenti" };

  function applyRemoteOps(ops, localDeviceId) {
    /* Applica le operazioni remote in locale con merge LWW.
       Ordine: wallet → category → transaction (per FK). */
    var byEntity = { wallet: [], category: [], transaction: [] };
    (ops || []).forEach(function (op) {
      var e = op.entity;
      if (byEntity[e]) byEntity[e].push(op);
    });

    var applied = 0, skipped = 0;
    var order = ["wallet", "category", "transaction"];

    // Promesse sequenziali per rispettare l'ordine
    var p = Promise.resolve();
    order.forEach(function (entity) {
      byEntity[entity].forEach(function (op) {
        p = p.then(function () {
          var store = ENTITY_STORE[entity];
          var uid = op.uid;
          var fields = op.fields || {};
          return DB.get(store, uid).then(function (local) {
            if (local) {
              // Record esiste: chi vince?
              if (!_wins(op.rev, op.updated_at, op.device_id,
                         local.rev, local.updated_at, localDeviceId)) {
                skipped++;
                return;
              }
            }
            // Remoto vince (o record nuovo): applica
            var obj = _applyFields(local, entity, fields);
            applied++;
            return DB.put(store, obj);
          });
        });
      });
    });

    return p.then(function () {
      return { applied: applied, skipped: skipped };
    });
  }

  // ── push: invia ops locali non pushate al PC ────────────────────────────
  function pushToPC(apiBase) {
    return Promise.all([getDeviceId(), DB.getUnpushedDiary()]).then(function (r) {
      var deviceId = r[0], ops = r[1];
      if (!ops.length) return { pushed: 0 };
      var body = { schema: 1, device_id: deviceId, ops: ops.map(function (o) {
        var copy = {}; for (var k in o) { if (k !== "_pushed" && k !== "id") copy[k] = o[k]; }
        return copy;
      }) };
      var ctrl = new AbortController();
      var to = setTimeout(function () { ctrl.abort(); }, 10000);
      return fetch(apiBase + "/api/finanze/ops", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
        signal: ctrl.signal
      }).then(function (res) {
        clearTimeout(to);
        if (!res.ok) throw new Error("push-http-" + res.status);
        return res.json();
      }).then(function (data) {
        if (!data.ok) throw new Error("push-rejected");
        // Marca le ops come pushate
        return DB.markDiaryPushed(ops.map(function (o) { return o.id; })).then(function () {
          return { pushed: ops.length, serverLines: data.diary_lines };
        });
      });
    });
  }

  // ── pull: scarica ops del PC e applicale ─────────────────────────────────
  function pullFromPC(apiBase) {
    return Promise.all([getDeviceId(), DB.getMeta("pc_diary_cursor")]).then(function (r) {
      var deviceId = r[0], cursor = r[1] || 0;
      var ctrl = new AbortController();
      var to = setTimeout(function () { ctrl.abort(); }, 10000);
      return fetch(apiBase + "/api/finanze/diary?since=" + cursor, {
        signal: ctrl.signal,
        headers: { Accept: "application/json" }
      }).then(function (res) {
        clearTimeout(to);
        if (!res.ok) throw new Error("pull-http-" + res.status);
        return res.json();
      }).then(function (data) {
        if (!data.ok) return { applied: 0, skipped: 0 };
        return applyRemoteOps(data.ops, deviceId).then(function (result) {
          // Aggiorna il cursore
          return DB.setMeta("pc_diary_cursor", data.total_lines).then(function () {
            return result;
          });
        });
      });
    });
  }

  // ── primo avvio: snapshot ────────────────────────────────────────────────
  function initialSync(apiBase) {
    return getDeviceId().then(function (deviceId) {
      var ctrl = new AbortController();
      var to = setTimeout(function () { ctrl.abort(); }, 15000);
      return fetch(apiBase + "/api/finanze/snapshot", {
        signal: ctrl.signal,
        headers: { Accept: "application/json" }
      }).then(function (res) {
        clearTimeout(to);
        if (!res.ok) throw new Error("snapshot-http-" + res.status);
        return res.json();
      }).then(function (snap) {
        // Lo snapshot è una lista di record: applicali come ops
        var ops = [];
        ["wallets", "categorie", "movimenti"].forEach(function (key) {
          var entity = { wallets: "wallet", categorie: "category", movimenti: "transaction" }[key];
          (snap[key] || []).forEach(function (fields) {
            ops.push({
              schema: 1, uid: fields.uid, entity: entity,
              op: fields.deleted ? "delete" : "upsert",
              fields: fields, rev: fields.rev || 1,
              updated_at: fields.updated_at || "",
              device_id: snap.device_id || ""
            });
          });
        });
        return applyRemoteOps(ops, deviceId);
      });
    });
  }

  // ── sync completa: push + pull ───────────────────────────────────────────
  function fullSync(apiBase) {
    if (!navigator.onLine) return Promise.resolve({ ok: false, reason: "offline" });
    return DB.isEmpty().then(function (empty) {
      if (empty) {
        // Primo avvio: scarica lo snapshot
        return initialSync(apiBase).then(function (result) {
          return DB.setMeta("last_sync", new Date().toISOString()).then(function () {
            return { ok: true, initial: true, applied: result.applied };
          });
        });
      }
      // Sync normale: push poi pull
      return pushToPC(apiBase).then(function (pushResult) {
        return pullFromPC(apiBase).then(function (pullResult) {
          return DB.setMeta("last_sync", new Date().toISOString()).then(function () {
            return {
              ok: true,
              pushed: pushResult.pushed,
              applied: pullResult.applied,
              skipped: pullResult.skipped
            };
          });
        });
      });
    }).catch(function (err) {
      return { ok: false, reason: err.message || "errore" };
    });
  }

  // ── export/import manuale ────────────────────────────────────────────────
  function exportBundle() {
    return Promise.all([
      DB.getAll("wallets"), DB.getAll("categorie"), DB.getAll("movimenti"),
      getDeviceId(), DB.getAllDiary()
    ]).then(function (r) {
      return {
        schema: 1, type: "bundle",
        snapshot: {
          schema: 1, device_id: r[3],
          ts: new Date().toISOString(),
          wallets: r[0], categorie: r[1], movimenti: r[2]
        },
        diary: r[4].map(function (o) {
          var copy = {}; for (var k in o) { if (k !== "_pushed" && k !== "id") copy[k] = o[k]; }
          return copy;
        })
      };
    });
  }

  function importBundle(data) {
    return getDeviceId().then(function (deviceId) {
      var ops = [];
      var snap = data.snapshot || data;
      ["wallets", "categorie", "movimenti"].forEach(function (key) {
        var entity = { wallets: "wallet", categorie: "category", movimenti: "transaction" }[key];
        (snap[key] || []).forEach(function (fields) {
          ops.push({
            schema: 1, uid: fields.uid, entity: entity,
            op: fields.deleted ? "delete" : "upsert",
            fields: fields, rev: fields.rev || 1,
            updated_at: fields.updated_at || "",
            device_id: snap.device_id || ""
          });
        });
      });
      // Poi il diario (se presente)
      (data.diary || []).forEach(function (op) { ops.push(op); });
      return applyRemoteOps(ops, deviceId);
    });
  }

  return {
    getDeviceId: getDeviceId,
    recordOp: recordOp,
    applyRemoteOps: applyRemoteOps,
    fullSync: fullSync,
    pushToPC: pushToPC,
    pullFromPC: pullFromPC,
    initialSync: initialSync,
    exportBundle: exportBundle,
    importBundle: importBundle
  };
})();

File: app/shared/sync.py
"""Motore di sincronizzazione multi-dispositivo (v2, Fase 4).

Protocollo stile Cashew: diario append-only per dispositivo, merge last-write-wins
per record. Il contratto è in docs/SYNC-PROTOCOL.md.

Flusso (locale → remoto):
1. Ogni scrittura locale (insert/update/soft-delete) registra un'operazione nel
   diario del dispositivo (file JSONL append-only in app/data/sync/).
2. Il diario viene esposto via API (`GET /api/finanze/diary`).
3. Un dispositivo remoto scarica le operazioni e le applica con merge LWW.

Flusso (remoto → locale):
1. Le operazioni remote arrivano via `POST /api/finanze/ops`.
2. Vengono applicate con le stesse regole di merge.
3. Durante l'import i metadati (rev/updated_at) NON vengono ri-timbrati: restano
   quelli del dispositivo sorgente.

Privacy: qui non passa nulla verso l'esterno. Il diario sta sul filesystem locale
(app/data/sync/) e le API girano solo su 127.0.0.1 (vedi shared/config.py).
"""
import json
import logging
import threading
import uuid
from contextlib import contextmanager
from datetime import datetime
from pathlib import Path

from sqlalchemy import event, select
from sqlalchemy.orm import Session

from shared.config import APP_DIR
from shared.db import SessionLocal
from shared import settings_store

log = logging.getLogger("mymoney.sync")

# ── configurazione ──────────────────────────────────────────────────────────
SYNC_DIR = APP_DIR / "data" / "sync"
SCHEMA_VERSION = 1

# ── flag thread-local per sopprimere la registrazione durante l'import ──────
_ctx = threading.local()


def _is_importing() -> bool:
    return getattr(_ctx, "importing", False)


@contextmanager
def importing():
    """Context manager: dentro questo blocco i before_flush / after_commit NON
    registrano nel diario e NON ri-timbrano rev/updated_at."""
    _ctx.importing = True
    try:
        yield
    finally:
        _ctx.importing = False


# ── device id ───────────────────────────────────────────────────────────────
def get_device_id() -> str:
    """ID stabile del dispositivo (generato al primo uso, poi persistito)."""
    did = settings_store.get_setting("sync_device_id", "")
    if not did:
        did = "pc_" + uuid.uuid4().hex[:12]
        settings_store.set_setting("sync_device_id", did)
    return did


# ── serializzazione entità → dict sync ──────────────────────────────────────
def _iso(dt):
    return dt.isoformat() if dt else None


def _wallet_to_fields(w) -> dict:
    return {
        "uid": w.uid, "nome": w.nome, "tipo": w.tipo,
        "saldo_iniziale": round(w.saldo_iniziale or 0.0, 2),
        "note": w.note or "", "ordine": w.ordine,
        "colore": w.colore or "", "archiviato": bool(w.archiviato),
        "deleted": bool(w.deleted),
        "rev": w.rev, "updated_at": _iso(w.updated_at),
    }


def _category_to_fields(c) -> dict:
    return {
        "uid": c.uid, "nome": c.nome, "kind": c.kind or "",
        "archiviato": bool(c.archiviato), "deleted": bool(c.deleted),
        "rev": c.rev, "updated_at": _iso(c.updated_at),
    }


def _transaction_to_fields(t, session) -> dict:
    """Campi sync di un movimento. Le FK (wallet_id, category_id) vengono
    risolte in uid del record referenziato (l'id interno non esce mai)."""
    from finance.models import Wallet, Category
    w_uid, wt_uid, cat_uid = None, None, None
    if t.wallet_id:
        w = session.get(Wallet, t.wallet_id)
        w_uid = w.uid if w else None
    if t.wallet_to_id:
        wt = session.get(Wallet, t.wallet_to_id)
        wt_uid = wt.uid if wt else None
    if t.category_id:
        cat = session.get(Category, t.category_id)
        cat_uid = cat.uid if cat else None
    return {
        "uid": t.uid, "tipo": t.tipo, "data": _iso(t.data),
        "importo": round(t.importo or 0.0, 2),
        "wallet_uid": w_uid, "wallet_to_uid": wt_uid,
        "categoria_uid": cat_uid,
        "descrizione": t.descrizione or "",
        "giro_id": t.giro_id or "", "giro_aperta": bool(t.giro_aperta),
        "importo_ricevuto": (round(t.importo_ricevuto, 2)
                             if t.importo_ricevuto is not None else None),
        "data_ricevuto": _iso(t.data_ricevuto),
        "controparte": t.controparte or "",
        "deleted": bool(t.deleted),
        "rev": t.rev, "updated_at": _iso(t.updated_at),
    }


def _obj_to_sync(obj, session) -> dict | None:
    """Converte un oggetto SQLAlchemy in un dict sync (None se non è sincronizzabile)."""
    from finance.models import Wallet, Category, Transaction
    if isinstance(obj, Wallet):
        entity, fields = "wallet", _wallet_to_fields(obj)
    elif isinstance(obj, Category):
        entity, fields = "category", _category_to_fields(obj)
    elif isinstance(obj, Transaction):
        entity, fields = "transaction", _transaction_to_fields(obj, session)
    else:
        return None
    op = "delete" if getattr(obj, "deleted", False) else "upsert"
    return {
        "schema": SCHEMA_VERSION,
        "uid": obj.uid,
        "entity": entity,
        "op": op,
        "fields": fields,
        "rev": obj.rev,
        "updated_at": _iso(obj.updated_at),
        "device_id": get_device_id(),
        "ts": datetime.now().isoformat(),
    }


# ── registrazione automatica nel diario ─────────────────────────────────────

def _diary_path() -> Path:
    SYNC_DIR.mkdir(parents=True, exist_ok=True)
    return SYNC_DIR / f"changes-{get_device_id()}.jsonl"


def _write_diary(ops: list[dict]) -> None:
    """Append delle operazioni al diario locale (una riga JSON per operazione)."""
    if not ops:
        return
    path = _diary_path()
    with open(path, "a", encoding="utf-8") as f:
        for op in ops:
            f.write(json.dumps(op, ensure_ascii=False, default=str) + "\n")


def _setup_diary_hooks():
    """Registra gli hook SQLAlchemy per la registrazione automatica nel diario.

    Viene chiamato UNA volta all'import del modulo. Gli hook:
    - before_flush (dopo quello di models.py): raccoglie le operazioni pendenti
    - after_commit: le scrive nel diario (solo se il commit ha successo)
    - after_soft_rollback: pulisce le operazioni pendenti se il commit fallisce
    """
    from finance.models import _MODELLI_SYNC

    @event.listens_for(Session, "before_flush")
    def _collect_sync_ops(session, flush_context, instances):
        """Raccoglie le operazioni da registrare nel diario. Gira DOPO il
        before_flush di models.py (che timbra uid/rev/updated_at), quindi i
        valori sono definitivi."""
        if _is_importing():
            return
        pending = []
        # Nuovi record
        for obj in session.new:
            if isinstance(obj, _MODELLI_SYNC):
                entry = _obj_to_sync(obj, session)
                if entry:
                    pending.append(entry)
        # Record modificati
        for obj in session.dirty:
            if isinstance(obj, _MODELLI_SYNC) and session.is_modified(obj, include_collections=False):
                entry = _obj_to_sync(obj, session)
                if entry:
                    pending.append(entry)
        if pending:
            session.info.setdefault("_sync_pending", []).extend(pending)

    @event.listens_for(Session, "after_commit")
    def _flush_diary(session):
        """Scrive le operazioni nel diario SOLO se il commit ha successo."""
        ops = session.info.pop("_sync_pending", [])
        if ops:
            _write_diary(ops)

    @event.listens_for(Session, "after_soft_rollback")
    def _discard_pending(session, previous_transaction):
        """Pulisce le operazioni pendenti se il commit fallisce."""
        session.info.pop("_sync_pending", None)


# ── lettura del diario ──────────────────────────────────────────────────────

def diary_lines_count() -> int:
    """Numero totale di righe nel diario locale."""
    path = _diary_path()
    if not path.exists():
        return 0
    with open(path, "r", encoding="utf-8") as f:
        return sum(1 for _ in f)


def export_diary(since_line: int = 0) -> list[dict]:
    """Ritorna le operazioni del diario locale dal cursore `since_line` in poi
    (0-indexed: 0 = tutte, N = dalla riga N esclusa)."""
    path = _diary_path()
    if not path.exists():
        return []
    ops = []
    with open(path, "r", encoding="utf-8") as f:
        for i, line in enumerate(f):
            if i < since_line:
                continue
            line = line.strip()
            if line:
                try:
                    ops.append(json.loads(line))
                except json.JSONDecodeError:
                    pass  # riga corrotta: skip, non bloccare la sync
    return ops


# ── merge LWW ───────────────────────────────────────────────────────────────

def _wins(remote_rev, remote_updated, remote_device,
          local_rev, local_updated, local_device) -> bool:
    """True se il record remoto vince sul locale (LWW per record).
    Ordinamento: (rev, updated_at, device_id) — vince il più alto."""
    if (remote_rev or 0) != (local_rev or 0):
        return (remote_rev or 0) > (local_rev or 0)
    r_up = str(remote_updated or "")
    l_up = str(local_updated or "")
    if r_up != l_up:
        return r_up > l_up
    return str(remote_device or "") > str(local_device or "")


def _parse_dt(s):
    """Parse ISO-8601 → datetime NAIVE (None se invalido). Tollera il suffisso
    'Z' (UTC) prodotto dal telefono e lo riporta in ora locale naive, così la
    colonna resta omogenea: mai un mix naive/aware che sporcherebbe i confronti
    e i calcoli di date altrove."""
    if not s:
        return None
    try:
        dt = datetime.fromisoformat(str(s).replace("Z", "+00:00"))
    except (TypeError, ValueError):
        return None
    if dt.tzinfo is not None:
        dt = dt.astimezone().replace(tzinfo=None)
    return dt


def _schema_troppo_nuovo(s) -> bool:
    """True se il messaggio è di uno schema più nuovo di quello che capiamo."""
    try:
        return int(s) > SCHEMA_VERSION
    except (TypeError, ValueError):
        return False   # schema assente/illeggibile = trattalo come 1 (legacy)


def import_ops(ops: list[dict], source_device_id: str = "") -> dict:
    """Applica una lista di operazioni remote con merge LWW.

    Ritorna {applied: int, skipped: int, errors: int}.
    Le operazioni vengono raggruppate per entità e applicate nell'ordine
    corretto: wallet → category → transaction (per risolvere le FK).
    """
    from finance.models import Wallet, Category, Transaction

    # Raggruppa per entità e ordina per priorità di applicazione
    by_entity = {"wallet": [], "category": [], "transaction": []}
    for op in ops:
        entity = op.get("entity", "")
        if entity in by_entity:
            by_entity[entity].append(op)

    applied = skipped = errors = future = 0

    with importing():
        with SessionLocal() as db:
            # Mappa uid→id per risolvere le FK dei movimenti
            uid_to_wallet_id = {w.uid: w.id for w in db.execute(select(Wallet)).scalars().all()}
            uid_to_cat_id = {c.uid: c.id for c in db.execute(select(Category)).scalars().all()}
            device_id = get_device_id()

            for entity in ("wallet", "category", "transaction"):
                for op in by_entity[entity]:
                    if _schema_troppo_nuovo(op.get("schema")):
                        future += 1
                        continue
                    try:
                        result = _apply_one(db, op, entity, device_id,
                                            uid_to_wallet_id, uid_to_cat_id)
                        if result == "applied":
                            applied += 1
                        else:
                            skipped += 1
                    except Exception:
                        errors += 1
                        # Non blocca la sync, ma lascia traccia per il debug
                        # (solo uid/entity: niente importi o descrizioni nei log).
                        log.warning("sync: op saltata uid=%s entity=%s",
                                    op.get("uid", ""), entity, exc_info=True)

            db.commit()

    if future > 0:
        settings_store.set_setting("sync_needs_update", "1")

    return {"applied": applied, "skipped": skipped, "errors": errors, "future": future}


def _apply_one(db, op, entity, local_device_id, uid_to_wallet_id, uid_to_cat_id) -> str:
    """Applica UNA operazione con merge LWW. Ritorna 'applied' o 'skipped'."""
    from finance.models import Wallet, Category, Transaction

    uid = op.get("uid", "")
    fields = op.get("fields", {})
    remote_rev = op.get("rev", 0)
    remote_updated = op.get("updated_at", "")
    remote_device = op.get("device_id", "")

    if not uid or not fields:
        return "skipped"

    # Trova il record locale per uid
    ModelClass = {"wallet": Wallet, "category": Category, "transaction": Transaction}[entity]
    local = db.execute(select(ModelClass).where(ModelClass.uid == uid)).scalar_one_or_none()

    if local is not None:
        # Record esiste: confronta per decidere chi vince
        if not _wins(remote_rev, remote_updated, remote_device,
                     local.rev, _iso(local.updated_at), local_device_id):
            return "skipped"
        # Il remoto vince: aggiorna i campi
        _set_fields(local, entity, fields, uid_to_wallet_id, uid_to_cat_id)
    else:
        # Record non esiste: inserisci
        obj = ModelClass()
        obj.uid = uid
        _set_fields(obj, entity, fields, uid_to_wallet_id, uid_to_cat_id)
        db.add(obj)
        db.flush()
        # Aggiorna le mappe per eventuali record successivi che referenziano questo
        if entity == "wallet":
            uid_to_wallet_id[uid] = obj.id
        elif entity == "category":
            uid_to_cat_id[uid] = obj.id

    return "applied"


def _set_fields(obj, entity, fields, uid_to_wallet_id, uid_to_cat_id):
    """Imposta i campi di un record dai valori sync (fields dict)."""
    # Metadati sync (sempre presenti)
    obj.rev = fields.get("rev", 1)
    obj.updated_at = _parse_dt(fields.get("updated_at"))
    obj.deleted = bool(fields.get("deleted", False))

    if entity == "wallet":
        obj.nome = fields.get("nome", "")
        obj.tipo = fields.get("tipo", "altro")
        obj.saldo_iniziale = fields.get("saldo_iniziale", 0.0)
        obj.note = fields.get("note", "")
        obj.ordine = fields.get("ordine", 0)
        obj.archiviato = bool(fields.get("archiviato", False))
        obj.colore = fields.get("colore", "")

    elif entity == "category":
        obj.nome = fields.get("nome", "")
        obj.kind = fields.get("kind", "")
        obj.archiviato = bool(fields.get("archiviato", False))

    elif entity == "transaction":
        obj.tipo = fields.get("tipo", "uscita")
        obj.data = _parse_dt(fields.get("data")) or datetime.now()
        obj.importo = fields.get("importo", 0.0)
        obj.descrizione = fields.get("descrizione", "")
        obj.giro_id = fields.get("giro_id", "")
        obj.giro_aperta = bool(fields.get("giro_aperta", False))
        obj.importo_ricevuto = fields.get("importo_ricevuto")
        obj.data_ricevuto = _parse_dt(fields.get("data_ricevuto"))
        obj.controparte = fields.get("controparte", "")
        # Risolvi FK: uid → id locale
        w_uid = fields.get("wallet_uid")
        obj.wallet_id = uid_to_wallet_id.get(w_uid) if w_uid else None
        wt_uid = fields.get("wallet_to_uid")
        obj.wallet_to_id = uid_to_wallet_id.get(wt_uid) if wt_uid else None
        cat_uid = fields.get("categoria_uid")
        obj.category_id = uid_to_cat_id.get(cat_uid) if cat_uid else None


# ── snapshot ────────────────────────────────────────────────────────────────

def build_snapshot() -> dict:
    """Fotografia completa di tutti i dati sync (wallet + categorie + movimenti).
    Usata per il primo avvio di un dispositivo nuovo."""
    from finance.models import Wallet, Category, Transaction
    with SessionLocal() as db:
        ws = list(db.execute(select(Wallet)).scalars().all())
        cs = list(db.execute(select(Category)).scalars().all())
        ts = list(db.execute(select(Transaction)).scalars().all())
        wallets = [_wallet_to_fields(w) for w in ws]
        categorie = [_category_to_fields(c) for c in cs]
        movimenti = [_transaction_to_fields(t, db) for t in ts]
    return {
        "schema": SCHEMA_VERSION,
        "device_id": get_device_id(),
        "ts": datetime.now().isoformat(),
        "diary_lines": diary_lines_count(),   # cursore iniziale per il client
        "wallets": wallets,
        "categorie": categorie,
        "movimenti": movimenti,
    }


def apply_snapshot(data: dict) -> dict:
    """Applica uno snapshot (primo avvio): tratta ogni record come un upsert."""
    if _schema_troppo_nuovo(data.get("schema")):
        settings_store.set_setting("sync_needs_update", "1")
        return {"applied": 0, "skipped": 0, "errors": 0, "future": 1}

    ops = []
    for entity, key in (("wallet", "wallets"), ("category", "categorie"),
                        ("transaction", "movimenti")):
        for fields in data.get(key, []):
            ops.append({
                "schema": SCHEMA_VERSION,
                "uid": fields.get("uid", ""),
                "entity": entity,
                "op": "delete" if fields.get("deleted") else "upsert",
                "fields": fields,
                "rev": fields.get("rev", 1),
                "updated_at": fields.get("updated_at", ""),
                "device_id": data.get("device_id", ""),
                "ts": data.get("ts", ""),
            })
    return import_ops(ops, source_device_id=data.get("device_id", ""))


# ── export/import file (fallback manuale) ───────────────────────────────────

def export_bundle() -> dict:
    """Bundle completo: snapshot + diario. Per backup o sync manuale."""
    return {
        "schema": SCHEMA_VERSION,
        "type": "bundle",
        "snapshot": build_snapshot(),
        "diary": export_diary(since_line=0),
    }


def import_bundle(data: dict) -> dict:
    """Importa un bundle (snapshot + diario) da un altro dispositivo."""
    result = {"applied": 0, "skipped": 0, "errors": 0, "future": 0}
    # Prima lo snapshot (crea i record base)
    snap = data.get("snapshot")
    if snap:
        r = apply_snapshot(snap)
        result["applied"] += r.get("applied", 0)
        result["skipped"] += r.get("skipped", 0)
        result["errors"] += r.get("errors", 0)
        result["future"] += r.get("future", 0)
    # Poi il diario (aggiorna con le modifiche più recenti)
    diary = data.get("diary", [])
    if diary:
        r = import_ops(diary)
        result["applied"] += r.get("applied", 0)
        result["skipped"] += r.get("skipped", 0)
        result["errors"] += r.get("errors", 0)
        result["future"] += r.get("future", 0)
    return result


# ── inizializzazione (chiamata all'import del modulo) ───────────────────────
_setup_diary_hooks()

